httpcache_test.go

```diff
@@ -595,6 +596,91 @@ func TestGetWithDoubleVary(t *testing.T) {
 	}
 }
 
+func TestKeepOnError(t *testing.T) {
+	// Create a new memory cache
+	cache := NewMemoryCache()
+	transport := NewTransport(cache)
+
+	// Create a test server that returns a 500 error
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Header.Get("Should-Fail") != "" {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		if r.Header.Get("Should-Sleep") != "" {
+			time.Sleep(1 * time.Second)
+		}
+		w.Header().Set("Etag", "123")
+		w.Header().Set("Vary", "X-Random")
+	}))
+	defer server.Close()
+
+	// Create a new request
+	req, _ := http.NewRequest("GET", server.URL, nil)
+	// Set the keep-on-error header
+	req.Header.Set("Cache-Control", "keep-on-error=true")
+	req.Header.Set("X-Random", "123")
+
+	{
+		// Perform the request using the transport
+		resp, err := transport.RoundTrip(req)
+		if err != nil {
+			t.Fatalf("Expected error, got nil")
+		}
+		defer resp.Body.Close()
+		// Check if the cache was kept
+		cachedResp, ok := cache.Get(cacheKey(req))
+		if !ok {
+			t.Fatalf("Expected cache to be kept, but it was not")
+		}
+		// Check if the cached response is valid
+		if cachedResp == nil {
+			t.Fatalf("Expected cached response, got nil")
+		}
+	}
+	// retry request with error from server
+	{
+		req.Header.Set("Should-Fail", "true")
+		req.Header.Set("If-Modified-Since", "Thu, 01 Jan 1970 00:00:00 GMT") // force revalidation
+		resp, err := transport.RoundTrip(req)
+		if err != nil {
+			t.Fatalf("Server returned error but transport returned nil: %v", err)
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != http.StatusInternalServerError {
+			t.Fatalf("response status code isn't 500 Internal Server Error: %v", resp.StatusCode)
+		}
+		cachedResp, ok := cache.Get(cacheKey(req))
+		if !ok {
+			t.Fatalf("Expected cache to be kept, but it was not")
+		}
+		// Check if the cached response is valid and not replaced by error response
+		if cachedResp == nil {
+			t.Fatalf("Expected cached response, got nil")
+		}
+	}
+	{
+		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+		defer cancel()
+		req, _ = http.NewRequestWithContext(ctx, "GET", server.URL, nil)
+		req.Header.Set("Should-Sleep", "true")
+		req.Header.Set("If-Modified-Since", "Thu, 01 Jan 1970 00:00:00 GMT") // force revalidation
+		req.Header.Set("X-Random", "123")
```

Comment:
qq: Want to make sure I'm reading this correctly - this header does not add to the test in anyway, right?

Reply:
`X-Random` is actually getting used as a key to store the value in httpcache. Notice that in server I am doing `Vary` on `X-Random`. So any request matching this header value will try to use same cached response.

Reply:
It does make sense, but I was looking at how cacheKey is built ([code ref](https://github.com/launchdarkly/streamer/blob/c07343bd4e8001e320cbe266b675b145234937c2/vendor/github.com/launchdarkly/httpcache/httpcache.go#L42-L58)) and it seems this httpCache does not include Vary in the cache key. 


```
	// Ideally we'd compute the cache key by taking into account the Vary headers.
	// Ugly hack for now -- add the authorization header to the cache key.
```

Reply:
We could use `Authorization` header here I guess.